File: Week 1/credit.c
#include <cs50.h>
#include <stdio.h>

int countDigits(long n);
int getId(long n, int digits);
int luhnValidation(long n, int digits);
void typeOfCard(int digits, int id);

int main(void)
{
    long n = get_long("Number: ");

    int digits = countDigits(n);

    int id = getId(n, digits);

    if (luhnValidation(n, digits))
    {
        typeOfCard(digits, id);
    }
    else
    {
        printf("INVALID\n");
    }
}

int countDigits(long n)
{
    int c = 0;

    while (n != 0)
    {
        c++;
        n = n / 10;
    }

    return c;
}

int getId(long n, int digits)
{
    long divider = 1;

    for (int i = 0; i < digits - 2; i++)
    {
        divider = divider * 10;
    }

    return n / divider;
}

int luhnValidation(long n, int digits)
{
    int sum = 0;

    long modDiv = 100; // module divider
    long div = 10;     // integer divider
    int mult = 100;    // comon multiplyer

    for (int i = 0; i < digits / 2; i++)
    {
        int value = 2 * ((n % modDiv) / div);

        if (value > 9)
        {
            value -= 9;
        }

        sum += value;

        modDiv = modDiv * mult;
        div = div * mult;
    }

    modDiv = 10;
    div = 1;
    int times;

    if (digits % 2 == 0)
    {
        times = digits / 2;
    }
    else
    {
        times = digits / 2 + 1;
    }

    for (int i = 0; i < times; i++)
    {
        sum += (n % modDiv) / div;

        modDiv = modDiv * mult;
        div = div * mult;
    }

    if (sum % 10 == 0)
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

void typeOfCard(int digits, int id)
{
    if ((digits == 15) && (id == 33 || id == 37))
    {
        printf("AMEX\n");
    }
    else if ((digits == 16) && (id >= 51 && id <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((digits == 13 || digits == 16) && (id >= 40 && id <= 49))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week 6/credit.py
from cs50 import get_int


def main():
    number = str(get_int("Number: "))
    if validation(number):
        print(card_type(number))
    else:
        print("INVALID")


def card_type(num):
    l = len(num)

    if (l == 13 or l == 16) and num[0] == "4":
        return "VISA"
    elif l == 15 and (num[0:2] == "34" or num[0:2] == "37"):
        return "AMEX"
    elif l == 16 and (int(num[0:2]) > 50 and int(num[0:2]) <= 55):
        return "MASTERCARD"
    else:
        return "INVALID"


def validation(num):
    sum = 0
    l = len(num)

    pair = True if l % 2 == 0 else False

    for i in range(l):
        n = int(num[i])

        if (pair):
            if i % 2 == 0:
                n *= 2
                sum += n if n < 10 else n - 9
            else:
                sum += n
        else:
            if i % 2 == 0:
                sum += n
            else:
                n *= 2
                sum += n if n < 10 else n - 9

    return True if sum % 10 == 0 else False


main()
